core/src/main/java/com/linecorp/armeria/server/HttpHeaderUtil.java

```diff
@@ -0,0 +1,186 @@
+        }
+        try {
+            return InetAddress.getByAddress(NetUtil.createByteArrayFromIpAddressString(addr.getHostName()));
+        } catch (UnknownHostException e) {
+            return stayUnresolvedIfFailed ? addr.getAddress() : null;
+        }
+    }
+
+    @VisibleForTesting
+    @Nullable
+    static InetAddress getFirstValidAddress(String headerValue,
+                                            Function<String, String> valueConverter,
+                                            Predicate<InetAddress> filter) {
+        if (Strings.isNullOrEmpty(headerValue)) {
+            return null;
+        }
+        for (String value : CSV_SPLITTER.split(headerValue)) {
+            final String v = valueConverter.apply(value);
+            if (Strings.isNullOrEmpty(v)) {
+                continue;
+            }
+            try {
+                final InetAddress addr = createInetAddress(v);
+                if (filter.test(addr)) {
+                    return addr;
+                }
+            } catch (UnknownHostException e) {
+                logger.debug("Failed to resolve hostname: {}", v, e);
+            }
+        }
+        return null;
+    }
+
+    private static InetAddress createInetAddress(String address) throws UnknownHostException {
+        // Remote quotes. e.g. "[2001:db8:cafe::17]:4711" => [2001:db8:cafe::17]:4711
+        final String addr = address.charAt(0) == '"' ? QUOTED_STRING_TRIMMER.trimFrom(address)
+                                                     : address;
+
+        if (NetUtil.isValidIpV4Address(addr) || NetUtil.isValidIpV6Address(addr)) {
+            return InetAddress.getByAddress(NetUtil.createByteArrayFromIpAddressString(addr));
+        }
+
+        if (addr.charAt(0) == '[') {
+            final int delim = addr.indexOf(']');
+            if (delim < 0) {
+                throw new UnknownHostException("Invalid IPv6 address format: " + addr);
+            }
+            // Remove the port part. e.g. [2001:db8:cafe::17]:4711 => [2001:db8:cafe::17]
+            final String withoutPort = addr.substring(0, delim + 1);
+            if (NetUtil.isValidIpV6Address(withoutPort)) {
+                return InetAddress.getByAddress(NetUtil.createByteArrayFromIpAddressString(withoutPort));
+            } else {
+                return InetAddress.getByName(withoutPort);
+            }
+        }
+
+        final int delim = addr.indexOf(':');
+        if (delim >= 0) {
+            // Remove the port part. e.g. 192.168.1.1:4711 => 192.168.1.1
+            final String withoutPort = addr.substring(0, delim);
+            if (NetUtil.isValidIpV4Address(withoutPort)) {
+                return InetAddress.getByAddress(NetUtil.createByteArrayFromIpAddressString(withoutPort));
+            } else {
+                return InetAddress.getByName(withoutPort);
+            }
+        }
+
+        // Try to lookup the host name.
+        return InetAddress.getByName(addr);
```

Comment:
Perhaps we should just reject the request or skip the header entry when received a non-IP address value? We obviously do not want to perform a blocking operation here.